activation/post_service.go

```diff
@@ -0,0 +1,166 @@
+		PostServiceCmd:  "/bin/service",
+		DataDir:         config.DefaultDataDir,
+		NodeAddress:     "http://127.0.0.1:9093",
+		PowDifficulty:   config.DefaultConfig().PowDifficulty,
+		PostServiceMode: "light",
+	}
+
+	return cfg
+}
+
+// MainnetPostServiceConfig returns the default config for mainnet.
+func MainnetPostServiceConfig() PostServiceConfig {
+	cfg := PostServiceConfig{
+		PostServiceCmd:  "/bin/service",
+		DataDir:         config.DefaultDataDir,
+		NodeAddress:     "http://127.0.0.1:9093",
+		PowDifficulty:   config.MainnetConfig().PowDifficulty,
+		PostServiceMode: "fast",
+	}
+
+	return cfg
+}
+
+type PostServiceConfig struct {
+	PostServiceCmd string `mapstructure:"post-opts-post-service"`
+
+	DataDir         string        `mapstructure:"post-opts-datadir"`
+	NodeAddress     string        `mapstructure:"post-opts-node-address"`
+	PowDifficulty   PowDifficulty `mapstructure:"post-opts-pow-difficulty"`
+	PostServiceMode string        `mapstructure:"post-opts-post-service-mode"`
+}
+
+// PostService manages a local post service.
+type PostService struct {
+	logger *zap.Logger
+
+	cmdMtx sync.Mutex         // protects concurrent access to cmd and stop.
+	cmd    *exec.Cmd          // cmd holds the post service command.
+	stop   context.CancelFunc // stops the running command.
+
+	eg errgroup.Group // eg manages post service goroutines.
+}
+
+// NewPostService returns a new post service.
+func NewPostService(logger *zap.Logger, opts PostServiceConfig) (*PostService, error) {
+	ctx, stop := context.WithCancel(context.Background())
+	cmd := exec.CommandContext(
+		ctx,
+		opts.PostServiceCmd,
+		"--dir", opts.DataDir,
+		"--address", opts.NodeAddress,
+		"--pow-difficulty", opts.PowDifficulty.String(),
+		"--randomx-mode", opts.PostServiceMode,
+	)
+	cmd.Dir = filepath.Dir(opts.PostServiceCmd)
+	pipe, err := cmd.StderrPipe()
+	if err != nil {
+		stop()
+		return nil, fmt.Errorf("setup stderr pipe for post service: %w", err)
+	}
+
+	ps := &PostService{
+		logger: logger,
+		cmd:    cmd,
+		stop:   stop,
+	}
+	ps.eg.Go(ps.captureCmdOutput(pipe))
+	if err := ps.cmd.Start(); err != nil {
+		pipe.Close()
+		stop()
+		return nil, fmt.Errorf("start post service: %w", err)
+	}
+	ps.eg.Go(ps.monitorCmd(opts))
+	return ps, nil
+}
+
+func (ps *PostService) Close() error {
+	ps.cmdMtx.Lock()
+	defer ps.cmdMtx.Unlock()
+
+	ps.stop()
+	ps.eg.Wait()
+	return nil
+}
+
+// captureCmdOutput returns a function that reads from the given pipe and logs the output.
+// it returns when the pipe is closed.
+func (ps *PostService) captureCmdOutput(pipe io.ReadCloser) func() error {
+	return func() error {
+		for {
+			buf := make([]byte, 1024)
+			n, err := pipe.Read(buf)
+			switch err {
+			case nil:
+			case io.EOF:
+				return nil
+			default:
+				ps.logger.Warn("read from post service pipe", zap.Error(err))
+				return nil
+			}
+
+			ps.logger.Debug(string(buf[:n]))
+		}
+	}
+}
+
+func (ps *PostService) monitorCmd(opts PostServiceConfig) func() error {
```

Comment:
I think that the code would be simpler and could avoid sharing memory (and mutexes) if this method was:
- just running a flat loop instead of recursion
- keeping the `cmd` local. AFAIR, you need `cmd` on the `PostService` struct only for testing. Perhaps this can be tested in other way? Or you could just set the `pid` in an `PostService::atomic.Value`?

Consider something like this:
```go
// PostService manages a local post service.
type PostService struct {
	logger *zap.Logger

	stop context.CancelFunc // stops the running command.

	eg errgroup.Group // eg manages post service supervisor
}

// NewPostService returns a new post service.
func NewPostService(logger *zap.Logger, opts PostServiceConfig) (*PostService, error) {
	ctx, stop := context.WithCancel(context.Background())

	ps := &PostService{
		logger: logger,
		stop:   stop,
	}
	ps.eg.Go(func() error { return ps.run(ctx, opts) })

	return ps, nil
}

func (ps *PostService) Close() error {
	ps.stop()
	ps.eg.Wait()
	return nil
}

// captureCmdOutput returns a function that reads from the given pipe and logs the output.
// it returns when the pipe is closed.
func (ps *PostService) captureCmdOutput(pipe io.ReadCloser) func() error {
	return func() error {
		for {
			buf := make([]byte, 1024)
			n, err := pipe.Read(buf)
			switch err {
			case nil:
			case io.EOF:
				return nil
			default:
				ps.logger.Warn("read from post service pipe", zap.Error(err))
				return nil
			}

			ps.logger.Debug(string(buf[:n]))
		}
	}
}

func (ps *PostService) run(ctx context.Context, opts PostServiceConfig) error {
	for {
		cmd := exec.CommandContext(
			ctx,
			opts.PostServiceCmd,
			"--dir", opts.DataDir,
			"--address", opts.NodeAddress,
			"--pow-difficulty", opts.PowDifficulty.String(),
			"--randomx-mode", opts.PostServiceMode,
		)
		cmd.Dir = filepath.Dir(opts.PostServiceCmd)
		pipe, err := cmd.StderrPipe()
		if err != nil {
			return fmt.Errorf("setup stderr pipe for post service: %w", err)
		}

		if cmd.Start(); err != nil {
			pipe.Close()
			return fmt.Errorf("start post service: %w", err)
		}
		events.EmitPostServiceStarted()

		var eg errgroup.Group
		eg.Go(ps.captureCmdOutput(pipe))
		err = cmd.Wait()
		ps.logger.Warn("post service exited", zap.Error(err))
		events.EmitPostServiceStopped()
		if err := ctx.Err(); err != nil {
                        if err := eg.Wait(); err != nil {
                                ps.logger.Warn("output reading goroutine failed", zap.Error(err))
                        }
			return ctx.Err()
		}
                if err := eg.Wait(); err != nil {
                        ps.logger.Warn("output reading goroutine failed", zap.Error(err))
                }
	}
}
```

Reply:
I ended up doing something quite similar to what you suggested; I even found a way to make it testable 🙂, although with the need for a new dependency.

While `os.FindProcess` and related functionality in the stdlib can do what I need on Linux and Mac, it fails to e.g. check for a running process on windows. Doing this with only the stdlib would be a pain, so I decided to use `github.com/shirou/gopsutil` for this.